# Auditoría Forense (Precios y Proveedores)

## Objetivo: El más crítico. Detectar si estamos pagando precios diferentes por lo mismo (Dispersión) y qué proveedores son caros.

In [7]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.io as pio
from sqlalchemy import create_engine

# Configuración visual y de renderizado
pd.options.display.float_format = '{:,.2f}'.format
pio.renderers.default = "iframe"  # Clave para que se vea en JupyterLab/Docker

# Conexión
# Usamos el host 'postgres' que ya validamos que funciona en tu red interna
engine = create_engine('postgresql+psycopg2://analista_medhos:Medhos2025!@postgres:5432/medhos_dw')

print("🚀 2.4 AUDITORÍA FORENSE: PRECIOS Y PROVEEDORES (Interactivo)")

# =============================================================================
# 1. CARGA DE DATOS
# =============================================================================
query = """
SELECT r.cod_insumo, r.descripcion_insumo, r.fecha_movimiento, 
       r.precio_unitario, r.nombre_proveedor, c.tipo_cambio_oficial_venta
FROM raw_movimientos_siga r
LEFT JOIN dim_cotizaciones c ON r.fecha_movimiento = c.fecha
WHERE r.tipo_archivo_detectado = 'ENTRADA'
AND r.precio_unitario > 0
"""
df = pd.read_sql(query, engine)

# Procesamiento y Dolarización
df['fecha_movimiento'] = pd.to_datetime(df['fecha_movimiento']) # Asegurar formato fecha
df['cotizacion_usd'] = df['tipo_cambio_oficial_venta'].ffill().bfill()
df['unitario_usd'] = df['precio_unitario'] / df['cotizacion_usd']
df['mes'] = df['fecha_movimiento'].dt.to_period('M')

# =============================================================================
# 2. ALERTA DE DISPERSIÓN (Mismo mes, Mismo producto, Distinto Precio)
# =============================================================================
print("\n🚨 --- 1. ALERTA DE DISPERSIÓN DE PRECIOS (Intra-mes) ---")
dispersión = df.groupby(['cod_insumo', 'descripcion_insumo', 'mes'])['unitario_usd'].agg(['min', 'max', 'mean', 'count'])
dispersión['variacion_pct'] = ((dispersión['max'] - dispersión['min']) / dispersión['min']) * 100

# Filtro: Variación > 50% en el mismo mes y más de 1 compra
sospechosos = dispersión[(dispersión['variacion_pct'] > 50) & (dispersión['count'] > 1)].sort_values('variacion_pct', ascending=False)

print(f"Se detectaron {len(sospechosos)} casos de dispersión extrema (>50% dif en el mismo mes).")
print("Top 10 Casos más graves:")
display(sospechosos.head(10))

# =============================================================================
# 3. INVESTIGACIÓN FORENSE DE PROVEEDORES
# =============================================================================
print("\n🕵️ --- 2. INVESTIGACIÓN DE PROVEEDORES ---")
# Precio promedio de mercado por insumo
precio_mercado = df.groupby('cod_insumo')['unitario_usd'].mean().rename('precio_promedio_mercado')
df = df.merge(precio_mercado, on='cod_insumo')

# Sobreprecio pagado vs promedio
df['sobreprecio_pct'] = ((df['unitario_usd'] - df['precio_promedio_mercado']) / df['precio_promedio_mercado']) * 100

ranking_prov = df.groupby('nombre_proveedor').agg(
    compras_cant=('cod_insumo', 'count'),
    sobreprecio_promedio=('sobreprecio_pct', 'mean')
)
ranking_prov = ranking_prov[ranking_prov['compras_cant'] > 10].sort_values('sobreprecio_promedio', ascending=False)

print("Top 10 Proveedores con precios sistemáticamente superiores al promedio:")
display(ranking_prov.head(10))

# =============================================================================
# 4. DISPERSIÓN VISUAL INTERACTIVA (PLOTLY)
# =============================================================================
print("\n📊 --- 3. DISPERSIÓN VISUAL (Top 5 Insumos Sospechosos) ---")

if len(sospechosos) > 0:
    # Seleccionamos los Top 5 códigos sospechosos
    top_sospechosos = sospechosos.reset_index()['cod_insumo'].head(5).tolist()
    df_plot = df[df['cod_insumo'].isin(top_sospechosos)].copy()
    
    # Convertir fecha a string para que se vea bien en el tooltip
    df_plot['fecha_str'] = df_plot['fecha_movimiento'].dt.strftime('%Y-%m-%d')

    fig = px.box(
        df_plot, 
        x="unitario_usd", 
        y="descripcion_insumo", 
        color="nombre_proveedor", # Cada proveedor un color distinto
        points="all", # Muestra cada factura individual como un punto
        hover_data=["fecha_str", "nombre_proveedor", "unitario_usd"], # Info al pasar el mouse
        title="Auditoría Forense: Variación de Precios por Proveedor (USD)",
        height=600
    )

    fig.update_layout(
        xaxis_title="Precio Unitario (USD)",
        yaxis_title="Insumo",
        template="plotly_white",
        legend_title="Proveedores"
    )
    
    fig.show()
else:
    print("✅ No hay casos sospechosos suficientes para graficar.")

🚀 2.4 AUDITORÍA FORENSE: PRECIOS Y PROVEEDORES (Interactivo)

🚨 --- 1. ALERTA DE DISPERSIÓN DE PRECIOS (Intra-mes) ---
Se detectaron 195 casos de dispersión extrema (>50% dif en el mismo mes).
Top 10 Casos más graves:


min    max   mean  count  \
cod_insumo descripcion_insumo             mes                                 
39830      Reactivo p/LDH UV método AUTOM 2020-06 0.05  17.87   5.99      3   
52133      Det. de Biotinidasa (unidad)   2022-04 3.36 685.46 344.41      2   
39830      Reactivo p/LDH UV método AUTOM 2021-05 0.11  21.85  10.98      2   
39837      Reactivo p/ Fosfatasa alcalina 2019-10 0.11  11.87   5.98      4   
39830      Reactivo p/LDH UV método AUTOM 2022-11 0.27  27.72   9.42      3   
                                          2022-10 0.29  28.46  14.37      2   
24987      Sonda                          2022-09 0.26  13.50   9.08      3   
57110      Ibuprofeno 400 mg IV ampolla   2022-12 0.39  15.36   7.88      2   
50673      Sonda p/intub.nasogastrica 3mm 2025-04 0.32  12.18   6.25      2   
56816      Gasa dobladilla.8 capas 10x10c 2020-04 0.07   2.14   0.76      3   

                                                   variacion_pct  
cod_insumo descripcion_insumo             mes                     
39830      Reactivo p/LDH UV método AUTOM 2020-06      35,024.82  
52133      Det. de Biotinidasa (unidad)   2022-04      20,283.42  
39830      Reactivo p/LDH UV método AUTOM 2021-05      19,283.86  
39837      Reactivo p/ Fosfatasa alcalina 2019-10      10,350.47  
39830      Reactivo p/LDH UV método AUTOM 2022-11      10,299.34  
                                          2022-10       9,802.54  
24987      Sonda                          2022-09       5,161.83  
57110      Ibuprofeno 400 mg IV ampolla   2022-12       3,830.13  
50673      Sonda p/intub.nasogastrica 3mm 2025-04       3,720.87  
56816      Gasa dobladilla.8 capas 10x10c 2020-04       2,921.86


🕵️ --- 2. INVESTIGACIÓN DE PROVEEDORES ---
Top 10 Proveedores con precios sistemáticamente superiores al promedio:


compras_cant  sobreprecio_promedio
nombre_proveedor                                        
RO BOT S.R.L.                   21                104.72
FARMACIA DEL PUENTE             26                 96.36
SOLMED S.A                      11                 84.96
SAN JUAN S.R.L.                 14                 62.85
OPTOVISION S.A.                 18                 26.35
ROSENBLAT EZEQUIEL D           182                 16.24
ROYMED S.A                      13                 15.90
XIMAX S.R.L.                    15                 14.94
BIOLATINA S.R.L.              2003                 13.80
ZANPER S.A.                    220                 10.89


📊 --- 3. DISPERSIÓN VISUAL (Top 5 Insumos Sospechosos) ---


In [3]:
import pandas as pd
import plotly.express as px
import plotly.io as pio
from sqlalchemy import create_engine

# Configuración de renderizado
pio.renderers.default = "iframe"

print("🚀 2.4 (ANEXO) ZOOM FORENSE 2025: DISPERSIÓN DE PRECIOS")

# 1. CONEXIÓN Y CARGA (Para asegurar que tenemos los datos frescos)
DB_USER = 'analista_medhos'
DB_PASS = 'Medhos2025!'
DB_HOST = 'postgres'
DB_NAME = 'medhos_dw'
engine = create_engine(f'postgresql+psycopg2://{DB_USER}:{DB_PASS}@{DB_HOST}:5432/{DB_NAME}')

# Carga de Compras
query = """
SELECT r.cod_insumo, r.descripcion_insumo, r.fecha_movimiento, 
       r.precio_unitario, r.nombre_proveedor, r.nro_factura, c.tipo_cambio_oficial_venta
FROM raw_movimientos_siga r
LEFT JOIN dim_cotizaciones c ON r.fecha_movimiento = c.fecha
WHERE r.tipo_archivo_detectado = 'ENTRADA'
AND r.precio_unitario > 0
"""
df = pd.read_sql(query, engine)

# --- CORRECCIÓN DEL ERROR ---
# Convertimos explícitamente a Datetime
df['fecha_movimiento'] = pd.to_datetime(df['fecha_movimiento'])

# 2. PROCESAMIENTO
# Dolarizar unitario
df['cotizacion_usd'] = df['tipo_cambio_oficial_venta'].ffill().bfill()
df['unitario_usd'] = df['precio_unitario'] / df['cotizacion_usd']

# 3. FILTRO 2025
df_2025 = df[df['fecha_movimiento'].dt.year == 2025].copy()
df_2025['mes_str'] = df_2025['fecha_movimiento'].dt.strftime('%Y-%m')

print(f"   -> Analizando {len(df_2025)} compras registradas en 2025.")

# 4. CÁLCULO DE DISPERSIÓN
stats = df_2025.groupby(['cod_insumo', 'descripcion_insumo', 'mes_str'])['unitario_usd'].agg(['min', 'max', 'count'])
stats['variacion_pct'] = ((stats['max'] - stats['min']) / stats['min']) * 100

# Filtro: >50% variación y más de 1 compra
sospechosos_2025 = stats[(stats['variacion_pct'] > 50) & (stats['count'] > 1)].sort_values('variacion_pct', ascending=False)

print(f"\n🚨 ALERTA 2025: Se detectaron {len(sospechosos_2025)} insumos con dispersión extrema.")

if len(sospechosos_2025) > 0:
    print("TOP 10 CASOS MÁS GRAVES:")
    display(sospechosos_2025.head(10))

    # 5. VISUALIZACIÓN (PLOTLY)
    top_codigos = sospechosos_2025.reset_index()['cod_insumo'].head(20).tolist()
    df_plot = df_2025[df_2025['cod_insumo'].isin(top_codigos)].copy()
    df_plot = df_plot.sort_values('descripcion_insumo')

    print("\n📊 --- GRÁFICO FORENSE INTERACTIVO ---")
    fig = px.box(
        df_plot, 
        x="unitario_usd", 
        y="descripcion_insumo", 
        color="descripcion_insumo", 
        points="all", 
        hover_data=["nombre_proveedor", "fecha_movimiento", "nro_factura"],
        title="Dispersión de Precios Unitarios 2025 (Top Casos Sospechosos)",
        height=800
    )
    
    fig.update_layout(
        xaxis_title="Precio Unitario (USD)",
        yaxis_title="",
        showlegend=False,
        template="plotly_white"
    )
    fig.show()
else:
    print("✅ No hay casos sospechosos en 2025.")

🚀 2.4 (ANEXO) ZOOM FORENSE 2025: DISPERSIÓN DE PRECIOS
   -> Analizando 5171 compras registradas en 2025.

🚨 ALERTA 2025: Se detectaron 34 insumos con dispersión extrema.
TOP 10 CASOS MÁS GRAVES:


min   max  count  \
cod_insumo descripcion_insumo             mes_str                     
50673      Sonda p/intub.nasogastrica 3mm 2025-04 0.32 12.18      2   
59189      Bolsa de reservorio de silicon 2025-08 0.57  7.32      2   
58049      Set para administrar sangre y  2025-06 0.67  2.46      2   
54929      Sutura quirurgica de nylon Nº5 2025-05 1.00  2.60      2   
39872      Reactivo p/anticore p/hepatiti 2025-04 3.20  7.59      3   
                                          2025-01 3.34  7.80      2   
59600      Colectores HPV (Cell collectio 2025-06 4.82  9.43      3   
59601      Cepillo Esteril p/toma muestra 2025-06 2.22  4.34      3   
54939      Sutura quirurgica de poliglact 2025-04 0.79  1.51      3   
39839      Reactivo p/trigliceridos AUTOM 2025-09 0.11  0.20      4   

                                                   variacion_pct  
cod_insumo descripcion_insumo             mes_str                 
50673      Sonda p/intub.nasogastrica 3mm 2025-04       3,720.87  
59189      Bolsa de reservorio de silicon 2025-08       1,188.07  
58049      Set para administrar sangre y  2025-06         264.22  
54929      Sutura quirurgica de nylon Nº5 2025-05         159.91  
39872      Reactivo p/anticore p/hepatiti 2025-04         136.95  
                                          2025-01         133.49  
59600      Colectores HPV (Cell collectio 2025-06          95.52  
59601      Cepillo Esteril p/toma muestra 2025-06          95.47  
54939      Sutura quirurgica de poliglact 2025-04          91.23  
39839      Reactivo p/trigliceridos AUTOM 2025-09          84.00


📊 --- GRÁFICO FORENSE INTERACTIVO ---


In [5]:
# =============================================================================
# 5. ESCENARIO DE AUDITORÍA: TOLERANCIA DEL 20% SOBRE EL MÍNIMO
# =============================================================================
print("\n👮 --- 5. AUDITORÍA ESTRICTA: ¿CUÁNTO SE PAGÓ POR ENCIMA DEL MÍNIMO (+20%)? ---")

# Usamos el dataframe df_sim que ya tiene los datos de 2025 cargados del paso anterior
# Si no está en memoria, descomenta las líneas de carga del bloque anterior.

# 1. Encontrar el PRECIO MÍNIMO conseguido por mes y producto (El "Benchmark")
benchmark = df_sim.groupby(['cod_insumo', 'mes_str'])['unitario_usd'].min().reset_index()
benchmark.rename(columns={'unitario_usd': 'precio_minimo_logrado'}, inplace=True)

# 2. Cruzar con las compras reales
df_audit = df_sim.merge(benchmark, on=['cod_insumo', 'mes_str'])

# 3. Definir el Techo Aceptable (Mínimo + 20%)
TOLERANCIA = 0.20 
df_audit['precio_techo'] = df_audit['precio_minimo_logrado'] * (1 + TOLERANCIA)

# 4. Calcular el "Exceso" pagado
# Si pagué 150 y el techo era 120, perdí 30. Si pagué 100, perdí 0.
df_audit['sobreprecio_unitario'] = df_audit['unitario_usd'] - df_audit['precio_techo']
df_audit['sobreprecio_unitario'] = df_audit['sobreprecio_unitario'].apply(lambda x: max(x, 0)) # Solo positivos

df_audit['dinero_desperdiciado'] = df_audit['sobreprecio_unitario'] * df_audit['cantidad']

# 5. Resultados
total_desperdicio = df_audit['dinero_desperdiciado'].sum()
gasto_total_2025 = df_audit['gasto_real'].sum()

print(f"📉 RESULTADOS AUDITORÍA 2025 (Tolerancia {TOLERANCIA*100}%)")
print(f"-------------------------------------------------------------")
print(f"Gasto Total Analizado:             USD {gasto_total_2025:,.2f}")
print(f"Dinero Pagado en Exceso (>20%):    USD {total_desperdicio:,.2f}")
print(f"-------------------------------------------------------------")
print(f"Porcentaje de Ineficiencia:        {(total_desperdicio/gasto_total_2025)*100:.2f}% del presupuesto")

# 6. Top 10 Insumos con mayor ineficiencia monetaria
print("\n🏆 TOP 10 INSUMOS CON MAYOR DESPERDICIO DE DINERO:")
ranking_ineficiencia = df_audit.groupby('descripcion_insumo')['dinero_desperdiciado'].sum().sort_values(ascending=False).head(10)
display(ranking_ineficiencia.to_frame(name="Dinero Perdido (USD)").style.format("${:,.2f}"))

# 7. Visualización de los Culpables (Proveedores)
top_item_ineficiente = ranking_ineficiencia.index[0]
print(f"\n🔎 ZOOM AL CASO #1: {top_item_ineficiente}")
print("¿Quién vendió barato y quién caro?")

caso_1 = df_audit[df_audit['descripcion_insumo'] == top_item_ineficiente].sort_values('unitario_usd')
display(caso_1[['fecha_movimiento', 'nombre_proveedor', 'cantidad', 'unitario_usd', 'precio_techo', 'dinero_desperdiciado']].style.format({
    'unitario_usd': '${:,.2f}',
    'precio_techo': '${:,.2f}',
    'dinero_desperdiciado': '${:,.2f}'
}))


👮 --- 5. AUDITORÍA ESTRICTA: ¿CUÁNTO SE PAGÓ POR ENCIMA DEL MÍNIMO (+20%)? ---
📉 RESULTADOS AUDITORÍA 2025 (Tolerancia 20.0%)
-------------------------------------------------------------
Gasto Total Analizado:             USD 7,060,212.91
Dinero Pagado en Exceso (>20%):    USD 19,832.33
-------------------------------------------------------------
Porcentaje de Ineficiencia:        0.28% del presupuesto

🏆 TOP 10 INSUMOS CON MAYOR DESPERDICIO DE DINERO:



🔎 ZOOM AL CASO #1: Tubo para suero c/5ml gel y ac
¿Quién vendió barato y quién caro?



Los resultados son mixtos: hay una buena gestión general del presupuesto macro (baja ineficiencia total), pero hay **casos puntuales gravísimos** de dispersión de precios y problemas de datos que debemos corregir.

### 📋 Resumen Ejecutivo: Hallazgos Clave 2025

#### 1. Ineficiencia Financiera Controlada (0.28%) ✅
*   De los **USD 7.06 Millones** analizados en 2025, el "dinero tirado a la basura" (pagado más de un 20% por encima del mínimo) fue de solo **USD 19,832**.
*   **Conclusión:** El sistema de compras general es estable. No hay un desfalco sistémico masivo, pero sí "micro-fugas" que suman.

#### 2. La "Sonda de la Discordia" (Error de Unidad de Medida) 🚨
*   El caso de la *Sonda p/intub.nasogastrica* muestra una variación del **3,720%** ($0.32 vs $12.18).
*   **Diagnóstico:** Esto **NO es fraude**, es un error de carga.
    *   Un proveedor facturó por **UNIDAD** ($0.32).
    *   Otro facturó por **CAJA** ($12.18).
*   **Acción:** Antes de llevar esto a un tablero gerencial, debemos limpiar estos casos o el promedio se romperá.

#### 3. El Caso "Biolatina vs Nipro" (Ineficiencia Real) 💸
*   En el *Tubo para suero*, detectamos una pérdida neta de **$3,200 USD** en junio.
*   **La Evidencia:**
    *   Compraste a **Nipro** a **$0.15**.
    *   En el mismo mes, le compraste a **Biolatina** a **$0.24** (un 60% más caro).
*   **Conclusión:** Aquí no hay error de unidad. Es el mismo producto. Compras no coordinó con Nipro para cubrir todo el pedido y tuvo que salir a comprar caro a Biolatina.

#### 4. Proveedores en la Mira 🕵️
*   **RO BOT S.R.L.** y **FARMACIA DEL PUENTE** aparecen con sobreprecios promedio del **100%** respecto al mercado.
*   Aunque sus volúmenes sean bajos, sistemáticamente cobran el doble que el resto. Son candidatos para la conexión y los primeros Dashboards.